decisions/DR009-Batch-method-result-types.md

```diff
@@ -0,0 +1,272 @@
+
+As shorthand for the different test cases, we define
+* _vector_ as the case that returns an array of results.
+* _callback_ as the case that executes a `std::function` callback on
+  each result.
+* _callbackToVector_ as the case that builds an array as callbacks
+  are executed and then iterates over it.
+* _callbackFnPtr_ as the case that uses function pointers, rather than
+  `std::function`, as the callback.
+* _callbackLargeCapture_ as the case that uses a `std::function`
+  callback holding a lambda with capture data such that the small-size
+  optimisation is unavailable.
+
+### Batch size = 1
+
+The following tables gives the results of the (likely) usual case of a
+batch size of 1 (i.e. only querying a single entity reference).
+
+Firstly, the primary results comparing the array return value solution
+versus the `std::function` callback solution is shown below.
+
+| numerator/denominator     | mean | std dev |
+|---------------------------|------|---------|
+| vector/callback           | 1.29 | 0.761   |
+| vector/callbackToVector   | 1.15 | 0.570   |
+| callbackToVector/callback | 1.24 | 0.745   |
+
+The secondary test results investigating the performance impact of
+function pointers and large lambda captures are presented in the
+following table.
+
+| numerator/denominator         | mean | std dev |
+|-------------------------------|------|---------|
+| callbackFnPtr/callback        | 1.09 | 0.600   |
+| callbackLargeCapture/callback | 1.40 | 0.758   |
+
+
+### Batch size = 100
+
+The following tables give the results using a larger batch size of 100.
+
+Firstly, as before, the primary results are shown below.
+
+| numerator/denominator     | mean | std dev |
+|---------------------------|------|---------|
+| vector/callback           | 1.44 | 0.230   |
+| vector/callbackToVector   | 1.08 | 0.192   |
+| callbackToVector/callback | 1.35 | 0.184   |
+
+And again, the secondary test results are shown below.
+
+| numerator/denominator         | mean | std dev |
+|-------------------------------|------|---------|
+| callbackFnPtr/callback        | 1.00 | 0.121   |
+| callbackLargeCapture/callback | 1.01 | 0.148   |
+
+
+### Conclusion
+
+Overall, it seems clear that a callback design is the most performant.
+
+With a small batch size, the standard deviation in timing ratios is so
+large that in practice there is little discernible difference in the
+options. However, it does appear that the callback solution has the edge
+slightly.
+
+Interestingly, using callbacks and appending to a vector on the caller
+side (_callbackToVector_) appears slightly more performant than simply
+returning a vector directly (_vector_). This could be due to compiler
+intricacies (e.g. moving the vector across the DSO boundary), but more
+investigation is required.
+
+However, with larger batches we start to see the performance
+characteristics of the chosen option dominate over external factors.
+Here, the callback solution(s) is the clear winner.
+
```

Comment:
Per my musings above, I'd find it compelling to have concurrent examples worked, in addition to single-threaded.

Reply:
@meshula we've added some notes on threading [here](https://github.com/OpenAssetIO/OpenAssetIO/pull/483/commits/04e25971225bb0c3ca0077074a39e36a2585cede). How does this fit into your thoughts more generally?

Reply:
Yes, thanks for laying out the strategy explicitly.